### PR TITLE
IC-1116 complete SP performance report schema

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

adds the required fields to generate the service provider performance report in full.

see https://docs.google.com/spreadsheets/d/1OYTTW-xG3djMsrp9dRRAQkEgXNq1uKoZkRD1ZSe7-OE/edit#gid=1131579949 for further explanations of these fields. 

## What is the intent behind these changes?

make the report useful to SPs.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ReferralTest.kt [hunk 1]
```diff
@@ -41,4 +44,30 @@ internal class ReferralTest {
...
+        actionPlans = mutableListOf(
+          actionPlanFactory.createSubmitted()
+        )
+      )
+
+      assertThat(referral.approvedActionPlan).isNull()
+    }
+
+    @Test
+    fun `approvedActionPlan returns most recent approved action plan`() {
```
Comment (mahalrmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ReferralTest.kt:61]: Are there a couple of test scenarios here, one for latest approved and then one for latest created when non approved

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/reporting/serviceprovider/performance/PerformanceReportProcessor.kt [hunk 2]
```diff
@@ -20,13 +25,50 @@ class PerformanceReportProcessor(
...
+      actionPlanSessionRepository.findAllByActionPlanId(actionPlan.id)
+        .flatMap { session -> session.appointments }
+        .filter { appointment ->
+          appointment.appointmentFeedbackSubmittedAt != null &&
+            listOf(Attended.LATE, Attended.YES).contains(appointment.attended)
+        }
+    }.orEmpty()
+  }
+
+  fun getFirstAttendedDeliveryAppointment(referral: Referral): Appointment? {
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/reporting/serviceprovider/performance/PerformanceReportProcessor.kt:71]: Do we need a test for this one?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReport.kt [hunk 3]
```diff
@@ -33,4 +34,13 @@ data class EndOfServiceReport(
   @CollectionTable(name = "end_of_service_report_outcome", joinColumns = [JoinColumn(name = "end_of_service_report_id")])
   @NotNull val outcomes: MutableSet<EndOfServiceReportOutcome> = mutableSetOf(),
 
-)
+) {
+  val achievementScore: Float
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/EndOfServiceReport.kt:38]: hmm, in the original query this wasn't the case. but i think it does make sense to add this provision.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ReferralTest.kt (hunk 1 above)
Comment (tomsmyers) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ReferralTest.kt:61]: the non-appoved action plans are covered by the other getter 'currentActionPlan', which does not filter on approval.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/reporting/serviceprovider/performance/PerformanceReportProcessor.kt (hunk 2 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/reporting/serviceprovider/performance/PerformanceReportProcessor.kt:71]: yeh thanks for the reminder.

**Review by mahalrmoj — APPROVED** [review]

**Pull request merged**